Labs 2,3,4,5 submission

xxmorwellr opened the pull request:
I tried to complete all the tasks, and I hope the pulled version is a syntax complete one. But I did encounter some errors and limitations when I wanted to execute the tasks. I hope you can review my code and give me a solution to error3, thank you!

xxmorwellr committed:
lab2

xxmorwellr commented:
Hej! I have fixed my errors in the lab session today, thank you!

nastaiza14 commented:
Hey! If you have updated your code, try uploading the new version to your repo, I'll take a look :)

xxmorwellr committed:
lab2 updated

xxmorwellr commented:
> Hey! If you have updated your code, try uploading the new version to your repo, I'll take a look :)

All right! I have updated my code in my branch now, Thank you:)

nastaiza14 commented:
Hey! The machine works, but there are important fixes to be made:
- The machine continues even if I say an element that is not in the grammar. For example, if I say that I want to meet with "Chocolate", it accepts and goes to the next question. We want to continue only when the user's request is in the grammar.
- When something that is not "yes" or "no" is replied to the question : "Will it take the whole day?", the machine stops, goes to idle. So, try taking care out of grammar examples there too.

Send me a message if you need help with this!

xxmorwellr committed:
lab2 fixed

xxmorwellr committed:
lab2 fixed

xxmorwellr commented:
> Hey! The machine works, but there are important fixes to be made:
> 
> * The machine continues even if I say an element that is not in the grammar. For example, if I say that I want to meet with "Chocolate", it accepts and goes to the next question. We want to continue only when the user's request is in the grammar.
> * When something that is not "yes" or "no" is replied to the question : "Will it take the whole day?", the machine stops, goes to idle. So, try taking care out of grammar examples there too.
> 
> Send me a message if you need help with this!

Hey,  thanks for your advice!

For the first bug, I have fixed it and given more tolerance for state transitions, just as my report said about _limitations_. To be specific,  if the answer is out of grammar or "NoReply" for 5 seconds, the machine will be transferred to the proper state. I also tuned the utterances to make the conversation more fluent.

For the second one, I tested it with a deny answer yesterday and it worked; I think the stop may be caused by failed recognition, which goes back to bug1. 

**In all, I have fixed my machine as much as I can and updated it now.** Hope you can enjoy it!

*For optimizing the code, I have some questions:

1. about **self-conversion**: For No-reply issue,  I utilized _Main.hist_ to realize a similar logic. While, in the child state node, when the guard condition meets with "I cannot understand you" and I want to re-enter the current state, it fails by targeting the current state to itself. So I added an intermediate state to solve the problem. Do you have any suggestions or ideas? 

2. For the end node _Done_, it's easy to understand the given grammar that by clicking the button, the machine can restart again, while it doesn't work in fact. 

3. Additionally, if I want to set the **limitation** of re-raise times for no-reply, how can I set the parameter? I may miss the detail in your lecture. Could you show me the example code again？ I think it may be helpful in the later project. 

Thank you!

nastaiza14 commented:
Nice work!!
We can talk about the questions in the lab on Friday!

xxmorwellr committed:
Labs 2,3 submission

xxmorwellr committed:
Labs 2,3 submission

xxmorwellr committed:
Labs 2,3,4 submission

xxmorwellr committed:
Add files via upload

xxmorwellr committed:
Labs 2,3,4,5 submission

xxmorwellr committed:
Labs 2,3,4,5 submission

xxmorwellr committed:
files merge

xxmorwellr committed:
Labs 2,3,4,5 submission

xxmorwellr committed:
Lab5 updated

nastaiza14 commented:
For lab 5, help cannot be obtained at every single step, but you kind of solve it with the "I don't understand" (out of grammar) state, which clarifies what the user can say at any point, and I think that's good!
